XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+// moment. Though we are currently investigating how to make this safe.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
+
+// post() is called just after the handle's callback has finished, and will be
+// called regardless whether the handle's callback threw. If the handle's
+// callback did throw then hasThrown will be true.
+function post(id, hasThrown) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPPARSER where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after post() has
+// completed. This is the only callback that does not receive the context as
+// "this". Because it is called while the associated class is being destructed
+// it is unsafe to operate on the object.
+function destroy(id) { }
+
+// Add a set of hooks to be called during the lifetime of asynchronous events.
+// createHook() returns an instance of the AsyncHook constructor that can
+// control how those hooks are used. For example, enabling or disabling their
+// execution.
+asyncHook = async_wrap.createHook({ init, pre, post, destroy });
+
+// Allow callbacks of this AsyncHook instance to fire. This is not an implicit
+// action after running createHook(), and must be explicitly run to being
+// executing callbacks.
+asyncHook.enable();
+
+// Disable listening for new asynchronous events. Though this will not prevent
+// callbacks from firing on asynchronous chains that have already run within
+// the scope of an enabled AsyncHook instance.
+asyncHook.disable();
+
+// Enable hooks for the current synchronous execution scope. This will ensure
+// the hooks are not in effect in case of multiple returns, or if an exception
+// is thrown.
+asyncHook.scope();
+```
+
+### `async_wrap`
+
+The object returned from `require('async_wrap')`.
+
+#### `async_wrap.createHook(hooks)`
+
+* `hooks` {Object}
+* Return: {AsyncHook}
+
+Create a new `AsyncHook` instance that controls a given set of hooks. The four
+optional hooks are `init`/`pre`/`post`/`destroy`.
+
+#### `async_wrap.Providers`
+
+List of all providers that may trigger the `init` callback.
+
+### Constructor: `AsyncHook`
+
+The `AsyncHook` constructor returns an instance that contains information about
+the callbacks that are to fire during specific asynchronous events in the
+lifetime of the event loop. The focal point of these calls centers around the
+lifetime of `AsyncWrap`. These callbacks will also be called to emulate the
+lifetime of handles and requests that do not fit this model. For example,
+`HTTPPARSER` instances are recycled to improve performance. So the `destroy()`
+callback would be called manually after a connection is done using it, just
```

Comment:
@AndreasMadsen Sorry, didn't see this. Yes, it will be manually called. When a resource is brought out of the pool it will be treated as if it were a new resource. All we're basically omitting is the memory allocation and class instantiation cost.